Automattic/jetpack
# wpcomsh: Marketplace Software Manager

[[ p9o2xV-4ju-p2 | Transfer Improvement: Single Post-Transfer SSH job ]] is implemented to stop making multiple calls to `woa_async_client_ssh_command` from post-transfer jobs due to their concurrency and SSH connection limit. This solution implements a single post-transfer job to send a command to the Atomic site with command flags that can be included by hooking into the `woa_transfer_wpcomsh_cli_flags_filter` filter.

To install the marketplace software the current implementation hooks into the `wpcomsh_woa_post_transfer` action to read the software details from the atomic persistent data. The details are used to generate the WP_CLI commands to install and verify the installation of the marketplace software.

Fixes  [WPCOMSH: Refactor marketplace_post_transfer_manage_software](https://github.com/Automattic/dotcom-forge/issues/8528)

## Proposed changes:

* Add hook callback to woa.php to check if the flag --install-marketplace-software is set to start the marketplace software installation.
* Implement WPCOM Marketplace Software library. The library contains Installers, software products, and a command helper.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* To prepare your environment, please go to pdt2If-1G3-p2
* As a prerequisite to testing this feature, follow the steps described in this diff  D156897-code 
* After following the steps in the diff, the WPCOM_MARKETPLACE_SOFTWARE atomic persistent data should be available on the Atomic site.
* Connect via SSH to the Atomic site
* Since the plugins and themes you purchased before moving the site to Atomic were installed by the implementation that is currently in production, deactivate and uninstall them. You can use `wp plugin deactivate [plugin] && wp plugin uninstall [plugin]` for unmanaged products. For managed products, deactivate and remove the symlink. `wp plugin deactivate && rm [PATH_TO_PRODUCT]`
* To install the software, run `wp wpcomsh post-transfer --install-marketplace-software`
* Check that there are no errors and all the plugins, themes, and their dependencies were installed.
